### ISSUE TITLE
How to restructure Project object for modern REDCap API

### ISSUE THREAD

**Issue by SlightlyUnorthodox** [redcap-tools/PyCap#65]
## New Things

Hello all,

So I've started going through the PyCap code and comparing it to the changes that have been made to the REDCap API post version 6.0.0. While most things shouldn't be too difficult to add in, there were two major changes that might call for restructuring the **Project** class.
1. In version 6.5.0 a "Export Project Information" API method was added that would make initializing a project object cleaner and more descriptive, it returns a large field of values with only the standard [token, content, format] input required.
   - This method returns the following:
   
   ```
   project_id, project_title, creation_time, production_time, in_production, project_language, purpose, purpose_other, project_notes, custom_record_label, secondary_unique_field, is_longitudinal, surveys_enabled, scheduling_enabled, record_autonumbering_enabled, randomization_enabled, ddp_enabled, project_irb_number, project_grant_number, project_pi_firstname, project_pi_lastname, display_today_now_button
   ```
   - This would deprecate a lot of the manual work used in PyCap to find things like **is_longitudinal**
2. In version 6.11.0 a "Create Project" API method was added that would allow for creating new projects directly in PyCap as long as the user has a 64-character supertoken.
   - All input parameters:
   
   ```
   project_title, purpose, purpose_other, project_notes, is_longitudinal, surveys_enabled, record_autonumbering_enabled
   ```
   - Returns a 32-character API token for continued access to project
### Problems

Here are some of the potential problems in adding the above methods.
- Both methods listed above used the same **content = 'project'** parameter when talking to the REDCap API, the distinguishing factor being the length of the API token.
- If changes were made to the **Project** class, there are 3 major version changes to account for and handle.
  - Post version 6.0.0 when **Export REDCap Version** was added
  - Post version 6.5.0 when **Export Project Information** was added
  - Post version 6.11.0 when **Create Project** was added
- Beyond this a number of new API methods and parameters were added throughout REDCap 6.X. which might encourage tackling things in a more adaptive, version-specific manner.
### Options

I will admit to not having prior experience working with REDCap, so it's quite likely I've overlooked some things, but here are the potential options as I see them. From there, I'll leave it to the community to decide what path makes the most sense when it comes to tackling these issues.
- Add **Export Project Information** and **Create Project** as fully separate methods under the **Project** object and handle version issues individually.
- Rewrite the **Project** object's initialization function such that it starts with a bit of logic to the extent of:

``` python
try:
  version_num = redcap.Get_Version( ) # attempt to get redcap version
except:
  log.warning("REDCAP < 6.0.0 not fully supported, use at own risk")
  version_num = 0.0.0
if version_num >= 6.11.0 and len(api_token) == 64:
  redcap.Create_Project( ) # Create new project in REDCap
else if version_num >= 6.5.0:
  redcap.Export_Project_Information( ) # initialize using modern api methods
else:
  # initialize using current methods
```
- Beyond that, I think there may be call to introduce a more complicated version checking structure to check for existence of API methods and the parameters available at each version. I haven't fleshed out quite how that would work yet, but I imagine you'd build up a dictionary of parameters for for each method with the version # as the key, which once parsed through would give you the available parameters at version x.

A lot of this boils down to how much the PyCap's base can afford to be changed, how much code is dependent on it in it's current form, and what versions of REDCap people are actually working with now, i.e. do we need to support REDCap before version 6.0.0?

Thoughts, comments, suggestions?

-Dax

**Comment by sburns** [comment]
I could see the create project method being implmenetd as a classmethod on `Project`, maybe being called like:

```python
details = {
    ...  # with keys mirroring the required input parameters
}  # 
new_project_token = Project.create(url, supertoken, details)
project = Project(url, new_project_token)
# or
project = Project.create(url, supertoken, details)
# now interact w/ the project as normal
```

The first might be better as you're probably going to want to store that new token somewhere. The second looks way nicer IMO (if you're creating a project, you'd expect the return value to be a project) and if you need the new API token, just access it at `project.token`.

I'd like to keep the constructor the same in that it connects to an existing REDCap project. I think creating a project through the API should be a very deliberate act with a different calling method.

> Beyond that, I think there may be call to introduce a more complicated version checking structure to check for existence of API methods and the parameters available at each version.

It would probably help to soften the arguments to the main methods and turn them into `**kwargs` and then be smart about adding defaults, removing unused keys, etc. based on the API version. That logic alone could turn into a whole new class (that we could write a ton of tests against because it wouldn't need to hit the network at all).

**Comment by SlightlyUnorthodox** [comment]
The concern I had with **Create Project** as a classmethod of ```Project``` (if not folded into \__init \__) is that you'd have to create a project from a previously initialized project object containing its own token and url, which seems awkward at best.

What if we created a separate class to the effect of ```ProjectManager``` which would allow for creation of new projects and management of initialized ```Project``` objects? Maybe something with some of the following functionality,
* ```ProjectManager.create_project( )```: create new ```Project```
* ```ProjectManager.view_projects( )```: view initialized ```Project```s
  * Potentially add a ```view_tokens( )``` component or build that into the data structure
* ```ProjectManager.store_tokens( )``` maybe add an api_token manager?

Maybe using an underlying table like so:

| Project     | Project Title | Source      | API token  |
| ------------- |:----------------|:-------------:| -----:|
| Project 1    | Title 1 | url 1 | token 1|
| Project 2      | Title 2 | url 2      |  token 2 |
| Source @ url 2 supertoken | NA | url 2      |  supertoken 2 |

I'm not sure exactly where to go with this, except that it seems the ```Project``` class is pretty self-contained. Maybe shoehorning in **Create_Project** doesn't actually make sense, particularly as having an interface for handling projects might allow for more thorough testing. 

At this point, I'm just throwing out ideas, I'll be working on adding methods to the ```Project``` class either way, and I agree on keeping the constructor the same, I'm hoping to maintain as much of the underlying structure as utterly possible.

**Comment by sburns** [comment]
> The concern I had with Create Project as a classmethod of Project (if not folded into __init __) is that you'd have to create a project from a previously initialized project object containing its own token and url, which seems awkward at best.

That's not how [classmethods](http://stackoverflow.com/questions/12179271/python-classmethod-and-staticmethod-for-beginner) work. Here's the pseudocode:

```python

class Project(object):

    @classmethod
    def create_project(cls, url, token, details, **constructor_kwargs):
        response = make API call to create project with url, supertoken & details
        return cls(url, response['new token'], **constructor_kwargs)
    
    def __init__(self, url, token, other stuff...):
         main constructor stuff
```

The `ProjectManager` idea is interesting. I'm terrified of storing tokens on behalf of users though. It's safer just to stuff them in well-designed environment variable names. Users on their own can stuff them in dot files but I wouldn't feel comfortable providing an official method for it, as there's a ton of PHI behind these tokens. Does that make sense?

**Comment by SlightlyUnorthodox** [comment]
Gotcha, for whatever reason I mentally parsed ```classmethod``` as "method of a class", rather than the python decorator. That makes complete sense.

As for storing tokens, I definitely understand. I'll stick with the implicit storage as environment variable that's already worked into the ```Project``` class.

Hopefully I'll have some of this pulled together for review in the next week or two.

**Comment by pwildenhain** [comment]
Since this was partially addressed in #106 and creating a project could potentially be out of scope, I'll close this out for now. If there is continued interest in creating a project through the API, then we can reopen.
